posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,631 @@
+
+    latest_created_at = sync_execute(SELECT_LATEST_CREATED_AT_QUERY.format(database=inputs.database))[0][0]
+
+    activity.logger.info("Creating JOIN table %s.%s", inputs.database, inputs.join_table_name)
+    sync_execute(CREATE_JOIN_TABLE_QUERY.format(database=inputs.database, table_name=inputs.join_table_name))
+
+    activity.logger.info("Populating JOIN table %s.%s", inputs.database, inputs.join_table_name)
+    sync_execute(
+        INSERT_INTO_JOIN_TABLE_QUERY.format(database=inputs.database, table_name=inputs.join_table_name),
+        {"latest_created_at": latest_created_at},
+    )
+
+    return latest_created_at.isoformat()
+
+
+@activity.defn
+async def select_persons_to_delete(inputs: QueryInputs) -> list[SerializablePersonOverrideToDelete]:
+    """Select the persons we'll override to lock them in and safely delete afterwards
+
+    New overrides may come in while we are executing this workflow, so we need to
+    preemptively select which persons we are going to override to ignore any new ones
+    coming in.
+
+    It's important that we only select those persons for which the first partition_id
+    in which they appear is covered by the current squash workflow. Otherwise, if there
+    is an override in an older partition that is not covered by the current workflow
+    we want to keep the override as there could still be events to squash.
+
+    The output of this activity is a dictionary that maps integer team_ids to sets of
+    person ids that are safe to delete. Team_id is used as a filter in later queries.
+    """
+    to_delete_rows = sync_execute(
+        SELECT_PERSONS_TO_DELETE_QUERY.format(database=inputs.database),
+        {"latest_created_at": inputs.latest_created_at},
+    )
+
+    if not isinstance(to_delete_rows, list):
+        # Could return None if no results or int if this were an insert.
+        # Mostly to appease type checker
+        return []
+
+    # We need to be absolutely sure which is the oldest event for a given person
+    # as we cannot delete persons that have events in the past that aren't being
+    # squashed by this workflow.
+    persons_to_delete = []
+    older_persons_to_delete = []
+    for row in to_delete_rows:
+        person_to_delete = PersonOverrideToDelete._make(row)
+        person_oldest_event_at = person_to_delete.oldest_event_at
+
+        try:
+            absolute_oldest_event_at = sync_execute(
+                SELECT_CREATED_AT_FOR_PERSON_EVENT_QUERY.format(database=inputs.database),
+                {
+                    "team_id": person_to_delete.team_id,
+                    "old_person_id": person_to_delete.old_person_id,
+                    "oldest_event_at": person_oldest_event_at,
+                },
+            )[0][0]
+
+        except IndexError:
+            # Let's be safe and treat this as no rows found.
+            absolute_oldest_event_at = EPOCH
+
+        # ClickHouse min() likes to return the epoch when no rows found.
```

Comment:
NULL-ish values on CH return some "empty" value based on the type of the column. i.e. `''` for strings, `0` for ints, and `1970-01-01` for datetimes